Allow rerunning individual examples in Xcode

bnickel opened the pull request:
## Problem statement

Today, if I click to run `FunctionalTests_ItSpec` in Xcode, I get the following **expected** output:

```
Test Suite 'Selected tests' started at 2022-09-02 12:41:16.743
Test Suite 'QuickTests.xctest' started at 2022-09-02 12:41:16.745
Test Suite 'FunctionalTests_ItSpec' started at 2022-09-02 12:41:16.745
Test Case '-[QuickTests.FunctionalTests_ItSpec _]' started.
Test Case '-[QuickTests.FunctionalTests_ItSpec _]' passed (0.002 seconds).
Test Case '-[QuickTests.FunctionalTests_ItSpec has_a_description_with_セレクター名に使えない文字が入っている___]' started.
Test Case '-[QuickTests.FunctionalTests_ItSpec has_a_description_with_セレクター名に使えない文字が入っている___]' passed (0.001 seconds).
Test Case '-[QuickTests.FunctionalTests_ItSpec when_an_example_has_a_unique_name__doesn_t_add_multiple_selectors_for_it]' started.
Test Case '-[QuickTests.FunctionalTests_ItSpec when_an_example_has_a_unique_name__doesn_t_add_multiple_selectors_for_it]' passed (0.001 seconds).
Test Case '-[QuickTests.FunctionalTests_ItSpec when_an_example_has_a_unique_name__has_a_unique_name]' started.
Test Case '-[QuickTests.FunctionalTests_ItSpec when_an_example_has_a_unique_name__has_a_unique_name]' passed (0.002 seconds).
Test Case '-[QuickTests.FunctionalTests_ItSpec when_two_examples_have_the_exact_name__has_exactly_the_same_name]' started.
Test Case '-[QuickTests.FunctionalTests_ItSpec when_two_examples_have_the_exact_name__has_exactly_the_same_name]' passed (0.001 seconds).
Test Case '-[QuickTests.FunctionalTests_ItSpec when_two_examples_have_the_exact_name__has_exactly_the_same_name_2]' started.
Test Case '-[QuickTests.FunctionalTests_ItSpec when_two_examples_have_the_exact_name__has_exactly_the_same_name_2]' passed (0.001 seconds).
Test Case '-[QuickTests.FunctionalTests_ItSpec when_two_examples_have_the_exact_name__makes_a_unique_name_for_each_of_the_above]' started.
Test Case '-[QuickTests.FunctionalTests_ItSpec when_two_examples_have_the_exact_name__makes_a_unique_name_for_each_of_the_above]' passed (0.001 seconds).
Test Suite 'FunctionalTests_ItSpec' passed at 2022-09-02 12:41:16.757.
	 Executed 7 tests, with 0 failures (0 unexpected) in 0.009 (0.012) seconds
Test Suite 'QuickTests.xctest' passed at 2022-09-02 12:41:16.765.
	 Executed 7 tests, with 0 failures (0 unexpected) in 0.009 (0.020) seconds
Test Suite 'Selected tests' passed at 2022-09-02 12:41:16.765.
	 Executed 7 tests, with 0 failures (0 unexpected) in 0.009 (0.022) seconds
Program ended with exit code: 0
```

Additionally, those 7 tests appear in the test sidebar.

If I try to run a single test from the sidebar, say `has_a_description_with_セレクター名に使えない文字が入っている___`, I get the following **unexpected** output:

```
Test Suite 'Selected tests' started at 2022-09-02 12:44:41.258
Test Suite 'QuickTests.xctest' started at 2022-09-02 12:44:41.259
Test Suite 'FunctionalTests_ItSpec' started at 2022-09-02 12:44:41.260
Test Suite 'FunctionalTests_ItSpec' passed at 2022-09-02 12:44:41.260.
	 Executed 0 tests, with 0 failures (0 unexpected) in 0.000 (0.000) seconds
Test Suite 'QuickTests.xctest' passed at 2022-09-02 12:44:41.260.
	 Executed 0 tests, with 0 failures (0 unexpected) in 0.000 (0.001) seconds
Test Suite 'Selected tests' passed at 2022-09-02 12:44:41.261.
	 Executed 0 tests, with 0 failures (0 unexpected) in 0.000 (0.002) seconds
Program ended with exit code: 0
```

The test is completely skipped without an explanation as to why, making it hard to impossible to rerun individual failing tests without rerunning a whole test class.

## Root cause

This issue is caused by the way `xctest` builds its `XCTestSuite` and the way `QuickSpec` generates the test methods used for execution.  Today, all that is facilitated by `QuickSpec.defaultTestSuite`, in the following sequence:

1. Either `xctest` identifies a QuickSpec subclass as a subclass of `XCTestCase` and adds it to the list of classes to test or the class to test is provided explicitly on invocation. 
2. `xctest` calls `QuickSpec.defaultTestSuite` on the class.
3. `QuickSpec.defaultTestSuite` calls `QuickConfiguration.configureSubclassesIfNeeded(world: World.sharedWorld)` to create the testing context if needed.
4. `QuickSpec.defaultTestSuite` calls `gatherExamplesIfNeeded()` to run `spec` and produce a collection of examples to execute.
5. `QuickSpec.defaultTestSuite` calls `super`.
6. `XCTestCase.defaultTestSuite` calls `_QuickSpecBase.testInvocations` to get the list of invocations.
7. `_QuickSpecBase.testInvocations ` calls `QuickSpec._qck_testMethodSelectors` to get the selectors to invoke.
8. `QuickSpec._qck_testMethodSelectors` uses `addInstanceMethod(for:classSelectorNames:)` to create a new Objective-C method that executes the block.
9. Functions return so that eventually the class has instance methods for each test and `xctest` has a test suite that can invoke each of them.

This breaks down in for individual test runs because they never call `defaultTestCase` on the class as a whole.  Instead they call `instancesRespond(to:)` with the invocation parameter.  This can be seen by looking at the back trace for both a class-level and an individual test run:

![Screen Shot 2022-09-01 at 8 21 20 PM](https://user-images.githubusercontent.com/662837/188229448-754e6c29-c14e-4a22-b5d8-8edccd0592c0.png)

![Screen Shot 2022-09-01 at 10 24 58 PM](https://user-images.githubusercontent.com/662837/188229427-f79dca13-9eaf-4e87-9905-72bcebc7d74a.png)

Both executions have a common root of `+[XCTestSuite prepareContainerSuite:toRunTestldentifiers:]` but one calls `defaultTestSuite` and the other builds it's test suite around an individual invocation with `+[XCTestSuite testSuiteForTestWithIdentifier:inTestClass:variations:]`, failing to produce a test if `instancesRespond(to:)` returns `false`.

It turns out that the function internally calls `+[XCTestSuite groupTestIdentifiers:intoClasses:andTestMethodIdentifiers:]` and handles the classes and methods in separate for-loops.

## Solution

The simplest solution, which is what I went for here, is just to implement `instancesRespond(to:)` and have it call `self.defaultTestSuite` before calling `super`.  This ensures the selector exists and the method returns `true`, yielding the expected results:

```
Test Suite 'Selected tests' started at 2022-09-02 13:09:29.337
Test Suite 'QuickTests.xctest' started at 2022-09-02 13:09:29.338
Test Suite 'FunctionalTests_ItSpec' started at 2022-09-02 13:09:29.338
Test Case '-[QuickTests.FunctionalTests_ItSpec has_a_description_with_セレクター名に使えない文字が入っている___]' started.
Test Case '-[QuickTests.FunctionalTests_ItSpec has_a_description_with_セレクター名に使えない文字が入っている___]' passed (0.002 seconds).
Test Suite 'FunctionalTests_ItSpec' passed at 2022-09-02 13:09:29.341.
	 Executed 1 test, with 0 failures (0 unexpected) in 0.002 (0.003) seconds
Test Suite 'QuickTests.xctest' passed at 2022-09-02 13:09:29.341.
	 Executed 1 test, with 0 failures (0 unexpected) in 0.002 (0.004) seconds
Test Suite 'Selected tests' passed at 2022-09-02 13:09:29.342.
	 Executed 1 test, with 0 failures (0 unexpected) in 0.002 (0.005) seconds
Program ended with exit code: 0
```

## Other considerations

`testInvocations` currently isn't cached in any way, so calling `self.defaultTestSuite` repeatedly clobbers the test implementations.  The function is idempotent, however, so this just results in extra work without breaking behavior.  I've also only ever seen `instancesRespond(to:)` execute once in testing.  If this is an issue, it can be mitigated by either caching `testInvocations` in an associated object or only invoking `self.defaultTestSuite` when `super.instancesRespond(to:)` is initially false.

I also considered if we should only inject the single method rather than all of them, but the current logic deals with name collisions at the time of insertion, so it would be a moderate refactor to pull out.

There are also a few other places this could live, like `+testCaseWithSelector:`.  I just picked the easiest to implement based on what I stumbled on first.

Lastly, I don't have any tests for this and I'm not really sure how to add one.

## Checklist

While not every PR needs it, new features should consider this list:

 - [ ] Does this have tests?
 - [ ] Does this have documentation?
 - [ ] Does this break the public API (Requires major version bump)?
 - [ ] Is this a new feature (Requires minor version bump)?

younata reviewed (commented):
This is very interesting, and kind of funny/sad in hindsight given the amount of effort many contributors have spent to get this feature working.

I wasn't able to get this to work until I modified `QuickSpec.m` to implement `+instancesRespondToSelector:` with the same implementation as this (the swift version wasn't getting called when I ran it locally. Xcode 14.0, macOS 12.5.1). Are you able to reproduce that?

bnickel commented:
> I wasn't able to get this to work until I modified `QuickSpec.m` to implement `+instancesRespondToSelector:` with the same implementation as this (the swift version wasn't getting called when I ran it locally. Xcode 14.0, macOS 12.5.1). Are you able to reproduce that?

I had been opening `Quick/` directly in Xcode rather than `Quick/Quick.xcworkspace`, so I was running tests using QuickSpec.swift instead of QuickSpec.m and didn't realize there was a second implementation.  I've added an Objective-C version of the method like you suggested.
